### PR TITLE
Try to fix branch coverage with async calls bug

### PR DESCRIPTION
**Pull request by MarcoRossignoli**
fixes https://github.com/tonerdo/coverlet/issues/113

The issue is related to compiler generated state machine:

```cs
...
TaskAwaiter awaiter;
if (num != 0)
{
    CoverageTracker.MarkExecuted(@"...\AppData\Local\Temp\CoverletAwaitBranchCoverage_04c15549-0ca8-4cd2-977b-01bd0a8ea52e", @"L,...\Git\CoverletAwaitBranchCoverage\CoverletAwaitBranchCoverage\Program.cs,9,9");
    CoverageTracker.MarkExecuted(@"...\AppData\Local\Temp\CoverletAwaitBranchCoverage_04c15549-0ca8-4cd2-977b-01bd0a8ea52e", @"L,...\Git\CoverletAwaitBranchCoverage\CoverletAwaitBranchCoverage\Program.cs,10,10");
    awaiter = Task.get_CompletedTask().GetAwaiter();
    if (!awaiter.IsCompleted)
    {
        CoverageTracker.MarkExecuted(@"...\AppData\Local\Temp\CoverletAwaitBranchCoverage_04c15549-0ca8-4cd2-977b-01bd0a8ea52e", @"B,...\Git\CoverletAwaitBranchCoverage\CoverletAwaitBranchCoverage\Program.cs,10,0");
        this.<>1__state = num = 0;
        this.<>u__1 = awaiter;
        Program.<Main>d__0 stateMachine = this;
        this.<>t__builder.AwaitUnsafeOnCompleted<TaskAwaiter, Program.<Main>d__0>(ref awaiter, ref stateMachine);
        return;
    }
}
...
```

If task complete synchronously this branch doens't hit.
I borrowed this idea from OpenCover.
I tested 2 strategies:
https://github.com/tonerdo/coverlet/pull/158/commits/3edf519e7ca578445f9c58fc020f5907f7528420#diff-9eb1afa9bd8141e48d456c120df924caR185 this one force hit on not covered branch
https://github.com/tonerdo/coverlet/pull/158/commits/06a5e59d06c14e7319cb88d80d9e02b74ab7c862#diff-9eb1afa9bd8141e48d456c120df924caR199 this one remove completely the branch if not hitted(but other one yes)

I prefer remove the branch...is more real than force hit. 
Let me know what do you think.

/cc @tonerdo

### CODE REVIEW

**Comment by tonerdo** [comment]
Hi @MarcoRossignoli, sorry for the delay in responding to this. Will take a look and get back to you

File: src/coverlet.core/Coverage.cs [hunk 1]
```diff
@@ -197,6 +196,31 @@ private void CalculateCoverage()
...
+                                if (moveNextBranch.Value.Method == branch.Value.Method && moveNextBranch.Value != branch.Value && moveNextBranch.Value.Hits == 0)
+                                {
+                                    branchesToRemove.Add(moveNextBranch);
+                                }
+                            }
+                        }
+                    }
+                    foreach (var branchToRemove in branchesToRemove)
+                    {
+                        document.Value.Branches.Remove(branchToRemove.Key);
```
Comment (MarcoRossignoli) [inline src/coverlet.core/Coverage.cs]: on next runtime version we can remove directly during enumeration(`Dictionary<(int Line, int Ordinal), Branch>`) https://github.com/dotnet/coreclr/pull/18854 https://github.com/dotnet/corefx/issues/31459

Comment (abe545) [reply src/coverlet.core/Coverage.cs]: I'm very interested in helping to get this pr merged - @MarcoRossignoli do you need to update this pr now that they released a new version of the clr last week?

Comment (MarcoRossignoli) [reply src/coverlet.core/Coverage.cs]: @abe545 thank's for the interest. I'll verify if above amend has been released, however it's not essential to this PR. When @tonerdo will be able to review we'll know if it's good to merge or not!

Comment (MarcoRossignoli) [reply src/coverlet.core/Coverage.cs]: NB. It's not my intention put pressure on maintainers!@tonerdo take all time you need, and thank's a lot for your work!

Comment (seesharper) [reply src/coverlet.core/Coverage.cs]: Would love to see this merged. Anything holding this back, @tonerdo ?

File: src/coverlet.core/Coverage.cs [hunk 2]
```diff
@@ -1,12 +1,11 @@
+using Coverlet.Core.Helpers;
+using Coverlet.Core.Instrumentation;
+using Coverlet.Core.Symbols;
```
Comment (tonerdo) [inline src/coverlet.core/Coverage.cs]: Let's move these imports below all `System.*` imports

**Comment by MarcoRossignoli** [comment]
> Apologies for the hold up on this

No problem.

Now result to this https://github.com/abe545/CoverletAwaitBranchCoverage/tree/master:

```
Starting test execution, please wait...

Total tests: 2. Passed: 2. Failed: 0. Skipped: 0.
Test Run Successful.
Test execution time: 4,6512 Seconds

Calculating coverage result...
  Generating report 'C:\Data\Source\Git\CoverletAwaitBranchCoverage\CoverletAwaitBranchCoverageTest\coverage.json'

+-----------------------------+--------+--------+--------+
| Module                      | Line   | Branch | Method |
+-----------------------------+--------+--------+--------+
| CoverletAwaitBranchCoverage | 100%   | 100%   | 100%   |
+-----------------------------+--------+--------+--------+


Total Line 100%
Total Branch 100%
Total Method 100%
```

**Comment by abe545** [comment]
@tonerdo no need to apologize for the delay, and thanks for all your hard work on this project!

**Comment by kevindqc** [comment]
Is this in `coverlet.msbuild` v2.4.0? I would assume so since this was done in October and 2.4.0 was published in November, but I get the uncovered branches when using await:
image

**Comment by MarcoRossignoli** [comment]
@kevindqc could you produce a simple repro?

**Comment by pape77** [comment]
I have a similar issue with this "ghost branches" which are never covered.
My code looks like this

```
    public class KeyProvider : IKeyProvider
    {
        private readonly IKeyRepository _repository;
        private readonly IMemoryCache _cache;

        public KeyProvider(IKeyRepository repository, IMemoryCache cache)
        {
            _repository = repository;
            _cache = cache;
        }

        public async Task<Key> GetByUuidAndTypeAsync(Guid uuid, KeyType type, CancellationToken cancellationToken)
        {
            return await _cache.GetOrCreateAsync($"uuid-{uuid.ToString()}-key-{type}",
                async entry => await _repository.FindOrCreateAsync(uuid, type, cancellationToken));
        }

   public class Key
    {
        public int Id { get; set; }

        public Guid Uuid { get; set; }

        [Required]
        public Guid Kid { get; set; }

        [Required]
        public KeyType KeyType { get; set; }
    }

 public enum KeyType
    {
        Cenc,
        Fairplay,
        Aes
    }
}
```
This code generates 4 branches, 2 of which I'm never able to cover

my tests:

```
  public class KeyProviderTests
    {
        private readonly Mock<IKeyRepository> _keyRepositoryMock;
        private readonly MemoryCache _memoryCache;
        private readonly Guid _testGuid = Guid.NewGuid();
        private readonly KeyProvider _keyProvider;

        public KeyProviderTests()
        {
            _memoryCache = new MemoryCache(new MemoryCacheOptions());

            _keyRepositoryMock = new Mock<IKeyRepository>();

            _keyProvider = new KeyProvider(_keyRepositoryMock.Object, _memoryCache);
        }

        [Theory(DisplayName = "Should call repository method with correct values")]
        [InlineData(KeyType.Cenc)]
        [InlineData(KeyType.Aes)]
        [InlineData(KeyType.Fairplay)]
        public async Task GetByUuidAndTypeAsync01(KeyType keyType)
        {
            _keyRepositoryMock.Setup(kr =>
                    kr.FindOrCreateAsync(_testGuid, keyType, It.IsAny<CancellationToken>()))
                .ReturnsAsync(new Key());

            await _keyProvider.GetByUuidAndTypeAsync(_testGuid, keyType, new CancellationToken());

            _keyRepositoryMock.Verify(
                kr => kr.FindOrCreateAsync(_testGuid, keyType, It.IsAny<CancellationToken>()), Times.Once);
        }

        [Theory(DisplayName = "Should get key from cache")]
        [InlineData(KeyType.Cenc)]
        [InlineData(KeyType.Aes)]
        [InlineData(KeyType.Fairplay)]
        public async Task GetByUuidAndTypeAsync02(KeyType keyType)
        {
            var cacheKey = $"uuid-{_testGuid.ToString()}-key-{keyType}";
            _memoryCache.Set(cacheKey, new Key());
           
            await _keyProvider.GetByUuidAndTypeAsync(_testGuid, keyType, new CancellationToken());

            _keyRepositoryMock.Verify(
                kr => kr.FindOrCreateAsync(_testGuid, keyType, It.IsAny<CancellationToken>()), Times.Never);
        }
}
```

The generated branches in my coverage.json are:

```
     "Keys.Application.Services.KeyProvider/<>c__DisplayClass3_0/<<GetByUuidAndTypeAsync>b__0>d": {
        "System.Void Keys.Application.Services.KeyProvider/<>c__DisplayClass3_0/<<GetByUuidAndTypeAsync>b__0>d::MoveNext()": {
          "Lines": {
            "27": 9
          },
          "Branches": [
            {
              "Line": 27,
              "Offset": 8,
              "EndOffset": 14,
              "Path": 0,
              "Ordinal": 0,
              "Hits": 3
            },
            {
              "Line": 27,
              "Offset": 81,
              "EndOffset": 83,
              "Path": 0,
              "Ordinal": 2,
              "Hits": 0
            },
            {
              "Line": 27,
              "Offset": 8,
              "EndOffset": 119,
              "Path": 1,
              "Ordinal": 1,
              "Hits": 0
            },
            {
              "Line": 27,
              "Offset": 81,
              "EndOffset": 147,
              "Path": 1,
              "Ordinal": 3,
              "Hits": 3
            }
          ]
        }
      },
```

**Comment by MarcoRossignoli** [comment]
@pape77 could create a full solution with you code?I see some references and would be great have buildable solution(sln) with simple 2 project similar to https://github.com/tonerdo/coverlet/issues/113#issuecomment-400723100, you can attach it here if you don't want to create a new repo.

**Comment by pape77** [comment]
[keys.zip](https://github.com/tonerdo/coverlet/files/2700537/keys.zip)

**Comment by MarcoRossignoli** [comment]
@pape77 thank's now I'm OOF bit I'll take a look ASAP

**Pull request merged**